File: levenshtein/julia/code.jl
# Calculates the Levenshtein distance between two strings using Wagner-Fischer algorithm
# Space Complexity: O(min(m,n)) - only uses two arrays instead of full matrix
# Time Complexity: O(m*n) where m and n are the lengths of the input strings
function levenshtein_distance(
    str1::String,
    str2::String,
    prev_row::Memory{Int32},
    curr_row::Memory{Int32},
)::Int
    # Early termination checks
    str1 == str2 && return 0
    (s1, s2) = (codeunits(str1), codeunits(str2))
    isempty(s1) && return length(s2)
    isempty(s2) && return length(s1)

    # Make s1 the shorter string for space optimization
    if length(s1) > length(s2)
        s1, s2 = s2, s1
    end

    m, n = length(s1), length(s2)

    # Initialize first row
    for i in 0:m
        @inbounds prev_row[i + 1] = i % Int32
    end

    # Main computation loop
    @inbounds for j in 1:n
        curr_row[1] = j % Int32

        for i in 1:m
            cost = s1[i] == s2[j] ? 0 : 1
            
            # Calculate minimum of three operations
            @inbounds curr_row[i + 1] = min(
                prev_row[i + 1] + 1,    # deletion
                curr_row[i] + 1,        # insertion
                prev_row[i] + cost      # substitution
            ) % Int32
        end

        # Swap rows
        prev_row, curr_row = curr_row, prev_row
    end

    return @inbounds prev_row[m + 1]
end


function main()
    if length(ARGS) < 2
        println("Please provide at least two strings as arguments.")
        exit(1)
    end

    min_distance = -1
    times = 0

    # Create arrays for reuse
    size = maximum(ncodeunits, ARGS; init=0) + 1
    v1 = Memory{Int32}(undef, size)
    v2 = Memory{Int32}(undef, size)

    # Compare all pairs of strings
    for i in 1:length(ARGS)
        for j in 1:length(ARGS)
            if i != j
                distance = levenshtein_distance(ARGS[i], ARGS[j], v1, v2)
                if min_distance == -1 || distance < min_distance
                    min_distance = distance
                end
                times += 1
            end
        end
    end

    println("times: ", times)
    println("min_distance: ", min_distance)
end

# Run main only if script is run directly
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: levenshtein/rust/code.rs
use std::env;

/// Calculates the Levenshtein distance between two strings using Wagner-Fischer algorithm
/// Space Complexity: O(min(m,n)) - only uses two rows instead of full matrix
/// Time Complexity: O(m*n) where m and n are the lengths of the input strings
fn levenshtein_distance(s1: &[u8], s2: &[u8]) -> u32 {
    let (s1_bytes, s2_bytes) = if s1.len() > s2.len() {
        (s2, s1)
    } else {
        (s1, s2)
    };

    let m = s1_bytes.len();

    // Use two rows instead of full matrix for space optimization
    let mut prev_row: Vec<u32> = (0..m + 1).map(|b| b as u32).collect();
    let mut curr_row: Vec<u32> = vec![0; m + 1];

    // Main computation loop
    for (j, s2) in s2_bytes.iter().enumerate() {
        curr_row[0] = (j + 1) as u32;

        for (i, s1) in s1_bytes.iter().enumerate() {
            let cost = if s1 == s2 { 0 } else { 1 };

            // Calculate minimum of three operations
            curr_row[i + 1] = (prev_row[i + 1] + 1) // deletion
                .min(curr_row[i] + 1) // insertion
                .min(prev_row[i] + cost); // substitution
        }

        // Swap rows
        std::mem::swap(&mut prev_row, &mut curr_row);
    }

    prev_row[m]
}

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();

    if args.len() < 2 {
        eprintln!("Please provide at least two strings as arguments.");
        return;
    }

    if !args.iter().all(|s| s.is_ascii()) {
        eprintln!("This benchmark assumes ASCII string values.");
        return;
    }

    if args.iter().any(|s| s.is_empty()) {
        eprintln!("This benchmark assumes non-empty strings.");
        return;
    }

    let mut min_distance = u32::MAX;
    let mut times = 0;

    // Compare all pairs of strings
    for (i, s1) in args.iter().enumerate() {
        for (j, s2) in args.iter().enumerate() {
            if i != j {
                let distance = levenshtein_distance(s1.as_bytes(), s2.as_bytes());
                min_distance = min_distance.min(distance);
                times += 1;
            }
        }
    }

    println!("times: {times}");
    println!("min_distance: {min_distance}");
}
